### PR TITLE
feat: move query performance polling to its own celery task in a performant manner

### PR DESCRIPTION
**Pull request by aspicer**
## Problem

As written, query performance polling blocks a user's polling request if clickhouse takes too long to return performance data
We're also doing redundant system.process queries to Clickhouse to get the data for each query.

## Changes

`poll_query_performance` is set up as a loop. Once the task ends, it schedules itself again, with a two second delay from when the last instance started. It uses redis as a lock, and if it detects that there's another instance of the loop running, it kills itself. This runs at most every 2 seconds or when the previous task finishes. If the clickhouse query is taking more than 10 seconds, the task aborts and starts again.

`start_poll_query_performance` is a simple scheduled task that makes sure an instance of `poll_query_performance` is running. It runs every 20 seconds, and if `poll_query_performance` hasn't run in 15 seconds, it puts a `poll_query_performance` task on the queue

This version extracts the query_id and the team_id from the clickhouse initial_query_id, which prevents it from having to try to infer team_id from redis keys.

This also adds a new run target for Pycharm that runs a threaded celery pool and separately runs the beat scheduler, which allows for more reliable local testing and development.

image

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Wrote tests for the behavior of task which controls restarting the poller if it dies.

Wrote tests to ensure the singleton behavior of the poller.

Wrote tests for the query status manager.

Added the ability to run the beat scheduler and a threadpool for celery in dev, and tested thoroughly 

**to test locally**
run the `Dev (Threaded Celery)` target in Pycharm

in `hogql/query.py` change line 182 to
`clickhouse_sql.replace("SELECT", "SELECT sleep(3), "),`

this will make the queries take a long time so you can see the loading bar locally. the queries will probably fail, but only after the bar shows up.

### CODE REVIEW

**Review by webjunkie — COMMENTED** [review]
Can we tackle this differently in that we just kick off the "query performance update" task when the user polls for the status each time?

This means it'll have a slight delay, but we don't introduce periodic polling and if nobody checks the status anymore then we also don't update the query status needlessly.
And it still solves the problem of the performance check blocking the request.

**Comment by aspicer** [comment]
@webjunkie
we could, absolutely. the first time they polled, we could kick it off, and then by the second time they polled, they'd have performance data.

we could even store a timestamp in redis and only kick off the job if it's been a few seconds since we last ran it, to avoid running this too many times

in the steady state though, i imagine that somebody would be running a query on our site pretty much around the clock - i think this ends up being pretty similar in load to constant polling. in fact under heavy site usage, this might be even more load, as we'd have to hit redis every time somebody polled. probably nothing compared to decide's usage of redis though, so maybe not worth designing around.

one benefit of this looping job comes in the way of future proofing - if we end up replacing the current query status polling system with a websocket or a long polling request, we'd have to then rearchitect this to work for that, vs the loop which wouldn't need to be changed.

**Comment by webjunkie** [comment]
@aspicer 
Thanks for the details. That makes sense then, if we think about a high volume of queries.
However, I'm wondering if we need this looping behavior that it starts itself again. Wouldn't it be enough to schedule it every few seconds, and make sure via the lock that it doesn't run multiple at the same time, and also use expirations and timeouts to make sure the tasks don't run late or too long?

**Comment by aspicer** [comment]
Yes, that would work. Do you have concerns about the looping behavior possibly causing issues

I like having the job schedule itself again, because it's the best way to guarantee a consistent running time. It can schedule itself to run exactly 2 seconds after it started.

Let's say instead we have the beat scheduler start the job every 2 seconds. The first thing the new job does is check Redis to see if the prior job has completed, or if a TIMEOUT amount of time has passed which might indicate that the previous job was killed.

This would be fine, but I think it introduces more variability and slightly more load into the running time than having the previous job schedule a new one.

I'm okay with changing it to do this if you're prefer, just trying to understand benefits. Thanks!

File: .run/Dev (Threaded Celery).run.xml
Comment (webjunkie) [inline .run/Dev (Threaded Celery).run.xml:1]: Thanks for remembering to adjust these run configs!

File: posthog/tasks/tasks.py [hunk 1]
```diff
@@ -509,6 +509,58 @@ def calculate_cohort() -> None:
...
+        if redis_client.get(POLL_SINGLETON_REDIS_KEY) != last_update:
+            logger.error("Poll query performance task terminating: another poller is running")
+            return
+        redis_client.set(POLL_SINGLETON_REDIS_KEY, start_time_str)
+        poll_query_performance_nontask()
+    except Exception as e:
+        logger.error("Poll query performance failed", error=e)
+
+    elapsed = time.time_ns() - start_time
+    if elapsed > 2e9:
```
Comment (webjunkie) [inline posthog/tasks/tasks.py:534]: We could put these magic numbers in descriptive variables.

File: .run/Celery Threads.run.xml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      <env name="REPLAY_EMBEDDINGS_ALLOWED_TEAM" value="1,2,3" />
+    </envs>
+    <option name="SDK_HOME" value="$PROJECT_DIR$/env/bin/python" />
+    <option name="WORKING_DIRECTORY" value="$PROJECT_DIR$/env/bin" />
+    <option name="IS_MODULE_SDK" value="true" />
+    <option name="ADD_CONTENT_ROOTS" value="true" />
+    <option name="ADD_SOURCE_ROOTS" value="true" />
+    <EXTENSION ID="PythonCoverageRunConfigurationExtension" runner="coverage.py" />
+    <option name="SCRIPT_NAME" value="$PROJECT_DIR$/env/bin/celery" />
+    <option name="PARAMETERS" value="-A posthog worker --without-heartbeat --without-mingle --loglevel=DEBUG --pool=threads" />
```
Comment (webjunkie) [inline .run/Celery Threads.run.xml:23]: Just FYI, I put `solo` solely to debug in Celery better. But it never occurred to me to test `threads` mode. 🤔 
If that works out in Pycharm, we actually don't need separate configs.

File: posthog/tasks/scheduled.py [hunk 3]
```diff
@@ -86,6 +87,8 @@ def setup_periodic_tasks(sender: Celery, **kwargs: Any) -> None:
     # Heartbeat every 10sec to make sure the worker is alive
     add_periodic_task_with_expiry(sender, 10, redis_heartbeat.s(), "10 sec heartbeat")
 
+    add_periodic_task_with_expiry(sender, 20, start_poll_query_performance.s(), "20 sec query performance heartbeat")
```
Comment (webjunkie) [inline posthog/tasks/scheduled.py:90]: Ah, I was suggesting expire, but we have that here already via `add_periodic_task_with_expiry`... not sure it's needed for the individual child tasks then.

**Review by webjunkie — APPROVED** [review]

**Pull request merged**